### PR TITLE
sis-scraper: introduce an additional 'phase 2'  pruning algorithm

### PR DESCRIPTION
**Pull request by johnnyapol**
This additional optimization phase reduces f20 and s21 down from 5 u64s
to just a single u64.

### CODE REVIEW

File: sis_scraper/main.py [hunk 1]
```diff
@@ -442,6 +443,51 @@ def optimize_column_ordering(data, num_columns=3):
...
+        # Now that we're on a (greatly) reduced working space, we can now prune using this
+        # less efficient algorithm
+        for index1 in range(BIT_VEC_SIZE):
+            # We want all (unordered) pairs of conflicting courses on the bit `index1`
+            pair_list = [pair for pair in combinations(sem_conflict_table[index1], 2)]
+
+            # This part essentially tries to see if some other bit(s) other than the current one will create a conflict
+            # for the conflicting classes in pair_list
+            # If there is, we can safely discard this bit.
+            for index2 in range(0, BIT_VEC_SIZE):
```
Comment (benjaminrsherman) [inline sis_scraper/main.py]: This should just be `range(BIT_VEC_SIZE)`

File: sis_scraper/main.py [hunk 2]
```diff
@@ -442,6 +443,51 @@ def optimize_column_ordering(data, num_columns=3):
...
+                # We can't use index1 to guarentee no conflicts, nor can we used other values that we determined to be
```
Comment (benjaminrsherman) [inline sis_scraper/main.py]: Typo in "guarantee" and "use"

File: sis_scraper/main.py [hunk 3]
```diff
@@ -442,6 +443,51 @@ def optimize_column_ordering(data, num_columns=3):
...
+                # unnecessary
+                if index2 in unnecessary_indices or index2 == index1:
```
Comment (benjaminrsherman) [inline sis_scraper/main.py]: Nitpick, but I'd flip these for both efficiency and to match the above comment

**Review by benjaminrsherman — COMMENTED** [review]
Assuming this optimization is correct, this is great!  I'm going to hold off on merging it until we can test that it provides the same conflict checking as the previous algorithm so conflict checking doesn't break (especially during registration time).  In the meantime, there are a few small changes to the code which I think would be helpful.

**Comment by johnnyapol** [comment]
I validated this using the sis_scraper_test.py that was recently introduced. The test is broken on tip (def dateranges_dont_overlap(year, ts1, ts2) does not work -- commenting out the call to this function allows it to pass on tip and when applied on my commit)

**Comment by johnnyapol** [comment]
I have introduced an additional commit to fix-up the test - it now passes on tip and when this change is applied.

**Comment by johnnyapol** [comment]
I've just pushed an additional patch to speed up pruning.

**Pull request closed without merging**